ParseUI-Widget/src/main/java/com/parse/widget/util/ParseQueryPager.java

```diff
@@ -0,0 +1,367 @@
+   *
+   * @param query The query for this {@code ParseQueryPager}.
+   */
+  public ParseQueryPager(ParseQuery<T> query) {
+    this(query, DEFAULT_PAGE_SIZE);
+  }
+
+  /**
+   * Constructs a new instance of {@code ParseQueryPager} with the specified query.
+   *
+   * @param query The query for this {@code ParseQueryPager}.
+   * @param pageSize The size of each page.
+   */
+  public ParseQueryPager(ParseQuery<T> query, int pageSize) {
+    this.query = new ParseQuery<>(query);
+    this.pageSize = pageSize;
+  }
+
+  /**
+   * @return the query for this {@code ParseQueryPager}.
+   */
+  public ParseQuery<T> getQuery() {
+    return query;
+  }
+
+  /**
+   * @return the size of each page for this {@code ParseQueryPager}.
+   */
+  public int getPageSize() {
+    return pageSize;
+  }
+
+  /**
+   * Returns the current page of the pager in the result set.
+   *
+   * The value is zero-based. When the row set is first returned the pager will be at position -1,
+   * which is before the first page.
+   *
+   * @return the current page.
+   */
+  public int getCurrentPage() {
+    synchronized (lock) {
+      return currentPage;
+    }
+  }
+
+  /**
+   * @return whether the pager has more pages.
+   */
+  public boolean hasNextPage() {
+    synchronized (lock) {
+      return hasNextPage;
+    }
+  }
+
+  /**
+   * @return whether the pager is currently loading the next page.
+   */
+  public boolean isLoadingNextPage() {
+    synchronized (lock) {
+      return loadNextPageTask != null && !loadNextPageTask.isCompleted();
+    }
+  }
+
+  /**
+   * @return the loaded objects.
+   */
+  public List<T> getObjects() {
+    return unmodifiableObjects;
+  }
+
+  public void addOnObjectsChangedCallback(OnObjectsChangedCallback callback) {
+    synchronized (lock) {
+      callbacks.add(callback);
+    }
+  }
+
+  public void removeOnObjectsChangedCallback(OnObjectsChangedCallback callback) {
+    synchronized (lock) {
+      callbacks.remove(callback);
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private void notifyRangeChanged(int positionStart, int positionEnd) {
+    synchronized (lock) {
+      for (OnObjectsChangedCallback callback : callbacks) {
+        callback.onItemRangeChanged(this, positionStart, positionEnd);
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private void notifyRangeInserted(int positionStart, int positionEnd) {
+    synchronized (lock) {
+      for (OnObjectsChangedCallback callback : callbacks) {
+        callback.onItemRangeInserted(this, positionStart, positionEnd);
+      }
+    }
+  }
+
+  private void setLoadNextPageTask(Task<List<T>> task) {
+    synchronized (lock) {
+      loadNextPageTask = task.continueWithTask(new Continuation<List<T>, Task<List<T>>>() {
+        @Override
+        public Task<List<T>> then(Task<List<T>> task) throws Exception {
+          synchronized (lock) {
+            loadNextPageTask = null;
```

Comment:
Why we need to clear `loadNextPageTask ` here, does this prevent us from calling `loadNextPageTask` multiple times?